### PR TITLE
Make `read!(::IO,::StridedArray)` call `unsafe_read` if possible

### PR DESCRIPTION
**Pull request by N5N3**
This PR makes `read!(s::IO,a::StridedArray)` call `unsafe_read`, if `a` has continous memory layout.
Since most of the checks are type-based, I think this does speed up, esspecially when `lock = true`.

BTW, the above dispatch for `read!(s::IO, a::Array{UInt8})` in L762 seems unnecessary, should we drop it?

### CODE REVIEW

**Review by vtjnash — APPROVED** [review]
SGTM. A similar change seems needed for `write` too, which currently appears to implement somewhat randomly unpredictable cases right now. There is also the question of padding, since `read` is usually not-implemented for most `T`, so whether this does `unsafe_read` or `read` is a different API.

File: base/io.jl [hunk 1]
```diff
@@ -872,7 +872,17 @@ function read!(s::IO, a::Array{UInt8})
...
+            sz::Dims = size(a)
+            st::Dims = strides(a)
+            msz, mst, n = merge_adjacent_dim(sz, st)
+            isbitdense = n == ndims(a) && mst == 1
+        end
+    end
+    if isbitdense
         GC.@preserve a unsafe_read(s, pointer(a), sizeof(a))
     else
         for i in eachindex(a)
```
Comment (vtjnash) [inline base/io.jl:888]:
```suggestion
        r = Ref{T}()
        for i in eachindex(a)
            unsafe_read(s, unsafe_convert(Ptr{T}, r), sizeof(r))
            a[i] = r[]
        end
```

File: base/io.jl [hunk 2]
```diff
@@ -872,10 +872,27 @@ function read!(s::IO, a::Array{UInt8})
...
-        for i in eachindex(a)
+        if isbitstype(T)
+            r = Ref{T}()
+            for i in eachindex(a)
+                @noinline unsafe_read(s, r, sizeof(r)) # r must be heap-allocated
+                a[i] = r[]
+            end
+        else
+            for i in eachindex(a)
             a[i] = read(s, T)
```
Comment (vtjnash) [inline base/io.jl:896]:
```suggestion
                a[i] = read(s, T)
            end
```

File: base/io.jl [hunk 3]
```diff
@@ -794,38 +794,45 @@ function write(s::IO, A::AbstractArray)
...
+function write(s::IO, A::StridedArray)
+    if !isbitstype(eltype(A))
         error("`write` is not supported on non-isbits arrays")
     end
-end
-
-function write(s::IO, a::SubArray{T,N,<:Array}) where {T,N}
-    if !isbitstype(T) || !isa(a, StridedArray)
-        return invoke(write, Tuple{IO, AbstractArray}, s, a)
+    elsz = sizeof(eltype(A))
```
Comment (vtjnash) [inline base/io.jl:809]: This is not the same as the zeroth stride of A, which is:
```suggestion
    elsz = elsize(typeof(A))
```
Though elsize seems to be implemented incorrectly for most AbstractArrays, so a better option is probably
```suggestion
    elsz = aligned_sizeof(T)
```

Comment (N5N3) [reply base/io.jl:809]: Looks like `elsize` is better for `Base`, as `aligned_sizeof` seems wrong for `ReinterpretArray`.

Comment (vtjnash) [reply base/io.jl]: The result for `ReinterpretArray` is wrong also, for which the correct result is `packedsize(T)`

Comment (N5N3) [reply base/io.jl]: Looks like `packedsize` is used only for bits `reinterpret`?
`ReinterpretArray` uses `sizeof` in `_getindex_ra` thus the padding won't be ignored.

Comment (vtjnash) [reply base/io.jl]: Oh weird. Alright then

File: base/io.jl [hunk 4]
```diff
@@ -866,20 +871,54 @@ end
 read(s::IO, ::Type{Bool}) = (read(s, UInt8) != 0)
 read(s::IO, ::Type{Ptr{T}}) where {T} = convert(Ptr{T}, read(s, UInt))
 
-function read!(s::IO, a::Array{UInt8})
-    GC.@preserve a unsafe_read(s, pointer(a), sizeof(a))
-    return a
+function read!(s::IO, A::AbstractArray{T}) where {T}
+    if isbitstype(T) && _checkcontiguous(Bool, A)
+        GC.@preserve A unsafe_read(s, pointer(A), elsize(A) * length(A))
```
Comment (N5N3) [inline base/io.jl:876]:
```suggestion
        GC.@preserve A unsafe_read(s, pointer(A), elsize(A) * length(A))
```

**Pull request merged**